appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java

```diff
@@ -311,667 +165,45 @@ private void createBuildMenu() {
-      db.show();
+    // TODO: This is probably wrong. Need to review how this functionality works with uibinder
+    // template.
+    if (!Ode.getInstance().hasSecondBuildserver()) {
+      buildDropDown.removeItemById(WIDGET_NAME_BUILD_ANDROID_APK2);
+      buildDropDown.removeItemById(WIDGET_NAME_BUILD_ANDROID_AAB2);
     }
-  }
-
-  private static class CompanionUpdateAction implements Command {
-    @Override
-    public void execute() {
-      DesignToolbar.DesignProject currentProject = Ode.getInstance().getDesignToolbar().getCurrentProject();
-      if (currentProject == null) {
-        Window.alert(MESSAGES.companionUpdateMustHaveProject());
-        return;
-      }
-      DesignToolbar.Screen screen = currentProject.screens.get(currentProject.currentScreen);
-      screen.blocksEditor.updateCompanion();
+    if (!AppInventorFeatures.hasYailGenerationOption()
+        || !Ode.getInstance().getUser().getIsAdmin()) {
+      buildDropDown.removeItemById(WIDGET_NAME_BUILD_YAIL);
     }
-  }
+    buildDropDown.removeUnneededSeparators();
 
-  private static class ShowSplashAction implements Command {
-    @Override
-    public void execute() {
-      Ode.getInstance().showWelcomeDialog();
+    if (!Ode.getUserAutoloadProject()) {
+      settingsDropDown.setItemHtmlById("AutoloadLastProject", MESSAGES.enableAutoload());
+      settingsDropDown.setCommandById("AutoloadLastProject", new EnableAutoloadAction());
     }
-  }
-
-  private static class WindowOpenAction implements Command {
-    private final String url;
-
-    WindowOpenAction(String url) {
-      this.url = url;
+    if (!Ode.getUserDyslexicFont()) {
+      settingsDropDown.setItemHtmlById("DyslexicFont", MESSAGES.enableOpenDyslexic());
+      settingsDropDown.setCommandById("DyslexicFont", new SetFontDyslexicAction());
     }
 
-    @Override
-    public void execute() {
-      Window.open(url, WINDOW_OPEN_LOCATION, WINDOW_OPEN_FEATURES);
-    }
-  }
+//    helpDropDown.removeUnneededSeparators();
 
-  private static class ImportComponentAction implements Command {
-    @Override
-    public void execute() {
-      new ComponentImportWizard().center();
+    if (!Ode.getInstance().getUser().getIsAdmin()) {
+      adminDropDown.removeFromParent();
     }
   }
 
-  private static class BuildComponentAction implements Command {
-    @Override
-    public void execute() {
-      // to be added
-    }
+  @UiFactory
+  public OdeMessages getMessages() {
+    return MESSAGES;
   }
 
-  private static class UploadComponentAction implements Command {
-    @Override
-    public void execute() {
-      new ComponentUploadWizard().show();
-    }
+  public void updateKeystoreStatus(boolean present) {
+    isKeystoreCached = true;
+    isKeystorePresent = present;
+    isKeystoreCheckPending = false;
+    fileDropDown.setItemEnabled(MESSAGES.deleteKeystoreMenuItem(), present);
+    fileDropDown.setItemEnabled(MESSAGES.downloadKeystoreMenuItem(), present);
```

Comment:
We should check that this correctly interacts with the readOnly status as seen in the constructor where these elements are disabled.

Reply:
I think this should be fine.